### PR TITLE
New UI

### PR DESCRIPTION
**Pull request by asafgabai**
- [ ] All [tests](https://github.com/jfrog/ide-plugins-common/actions/workflows/test.yml) passed. If this feature is not already covered by the tests, I added new tests.
-----
Depends on:
https://github.com/jfrog/build-info/pull/688
https://github.com/jfrog/ide-plugins-common/pull/93

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -32,7 +32,7 @@ dependencies {
     implementation(group: 'org.apache.httpcomponents', name: 'httpclient', version: '4.5.13') {
         exclude group: 'commons-codec', module: 'commons-codec'
     }
-    implementation group: 'com.jfrog.xray.client', name: 'xray-client-java', version: '0.11.0'
+    implementation group: 'com.jfrog.xray.client', name: 'xray-client-java', version: '0.13.0'
```
Comment (asafgabai) [inline build.gradle:35]: Reminder for myself to upgrade build-info version above before the merge.

File: src/main/java/com/jfrog/ide/common/tree/GeneralInfo.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+ */
+@JsonInclude(JsonInclude.Include.NON_EMPTY)
+public class GeneralInfo implements Serializable {
+
+    private static final long serialVersionUID = 1L;
+
+    private String componentId = "";
+    private String pkgType = "";
+    private String path = "";
+    private String sha1 = "";
```
Comment (yahavi) [inline src/main/java/com/jfrog/ide/common/tree/GeneralInfo.java:19]: I'm not sure we need the sha1 anymore
```suggestion
```

File: src/main/java/com/jfrog/ide/common/tree/DescriptorFileTreeNode.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
+package com.jfrog.ide.common.tree;
+
+import java.util.Collection;
+
+public class DescriptorFileTreeNode extends FileTreeNode {
```
Comment (yahavi) [inline src/main/java/com/jfrog/ide/common/tree/DescriptorFileTreeNode.java:5]:
```suggestion
@SuppressWarnings("unused")
public class DescriptorFileTreeNode extends FileTreeNode {
```

File: src/main/java/com/jfrog/ide/common/tree/DescriptorFileTreeNode.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        super(filePath);
+    }
+
+    public void addDependency(Artifact dependency) {
+        Artifact clonedDep = (Artifact) dependency.clone();
+        add(clonedDep);
+        if (clonedDep.getTopSeverity().isHigherThan(topSeverity)) {
+            topSeverity = clonedDep.getTopSeverity();
+        }
+        sortChildren();
```
Comment (yahavi) [inline src/main/java/com/jfrog/ide/common/tree/DescriptorFileTreeNode.java:16]: Removes code duplication
```suggestion
        addDependencies(Collections.singletonList(dependency));
```

File: src/main/java/com/jfrog/ide/common/tree/DescriptorFileTreeNode.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    }
+
+    public void addDependencies(Collection<Artifact> dependencies) {
+        for (Artifact dependency : dependencies) {
+            add(dependency);
```
Comment (yahavi) [inline src/main/java/com/jfrog/ide/common/tree/DescriptorFileTreeNode.java:21]: Shouldn't need to clone the artifact here?

File: src/main/java/com/jfrog/ide/common/tree/DescriptorFileTreeNode.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
+package com.jfrog.ide.common.tree;
+
+import java.util.Collection;
+
+public class DescriptorFileTreeNode extends FileTreeNode {
+    public DescriptorFileTreeNode(String filePath) {
+        super(filePath);
+    }
+
+    public void addDependency(Artifact dependency) {
```
Comment (yahavi) [inline src/main/java/com/jfrog/ide/common/tree/DescriptorFileTreeNode.java:10]: Is it a dependency or an artifact?

File: src/main/java/com/jfrog/ide/common/tree/DescriptorFileTreeNode.java [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+            add(dependency);
+            if (dependency.getTopSeverity().isHigherThan(topSeverity)) {
+                topSeverity = dependency.getTopSeverity();
+            }
+        }
+        sortChildren();
+    }
+
+    private void sortChildren() {
+        if (children != null && children.size() > 1) {
```
Comment (yahavi) [inline src/main/java/com/jfrog/ide/common/tree/DescriptorFileTreeNode.java:30]:
```suggestion
        if (CollectionUtils.isNotEmpty(children)) {
```

File: src/main/java/com/jfrog/ide/common/tree/Issue.java [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+import com.fasterxml.jackson.annotation.JsonInclude;
+import org.apache.commons.lang3.StringUtils;
+
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * @author yahavi
+ */
+@JsonInclude(JsonInclude.Include.NON_EMPTY)
```
Comment (yahavi) [inline src/main/java/com/jfrog/ide/common/tree/Issue.java:13]:
```suggestion
```

File: src/main/java/com/jfrog/ide/common/tree/Issue.java [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+public class Issue extends VulnerabilityOrViolation {
```
Comment (yahavi) [inline src/main/java/com/jfrog/ide/common/tree/Issue.java:14]:
```suggestion
public class IssueNode extends VulnerabilityOrViolation {
```

File: src/main/java/com/jfrog/ide/common/tree/Issue.java [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+    public List<String> getFixedVersions() {
+        return fixedVersions;
+    }
+
+    @SuppressWarnings("unused")
+    public void setFixedVersions(List<String> fixedVersions) {
+        this.fixedVersions = fixedVersions;
+    }
+
+    public List<String> getInfectedVersions() {
```
Comment (yahavi) [inline src/main/java/com/jfrog/ide/common/tree/Issue.java:68]: There are more unused in this class
```suggestion
    @SuppressWarnings("unused")
    public List<String> getInfectedVersions() {
```

File: src/main/java/com/jfrog/ide/common/tree/License.java [hunk 11]
```diff
@@ -0,0 +1,19 @@
+package com.jfrog.ide.common.tree;
+
+public class License {
+    private String name;
+    private String moreInfoUrl;
```
Comment (yahavi) [inline src/main/java/com/jfrog/ide/common/tree/License.java:5]:
```suggestion
    private final String name;
    private final String moreInfoUrl;
```

File: src/main/java/com/jfrog/ide/common/tree/LicenseViolation.java [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+import org.apache.commons.lang3.StringUtils;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * @author yahavi
+ */
+@JsonInclude(JsonInclude.Include.NON_EMPTY)
```
Comment (yahavi) [inline src/main/java/com/jfrog/ide/common/tree/LicenseViolation.java:14]:
```suggestion
```

File: src/main/java/com/jfrog/ide/common/tree/LicenseViolation.java [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+public class LicenseViolation extends VulnerabilityOrViolation {
```
Comment (yahavi) [inline src/main/java/com/jfrog/ide/common/tree/LicenseViolation.java:15]:
```suggestion
public class LicenseViolationNode extends VulnerabilityOrViolation {
```

File: src/main/java/com/jfrog/ide/common/tree/ResearchInfo.java [hunk 14]
```diff
@@ -0,0 +1,42 @@
+package com.jfrog.ide.common.tree;
+
+public class ResearchInfo {
+    private Severity severity;
+    private String shortDescription;
+    private String fullDescription;
+    private String remediation;
+    private SeverityReason[] severityReasons;
```
Comment (yahavi) [inline src/main/java/com/jfrog/ide/common/tree/ResearchInfo.java:8]: Let's make them final to get rid of compiler warnings

File: src/main/java/com/jfrog/ide/common/tree/SeverityReason.java [hunk 15]
```diff
@@ -0,0 +1,28 @@
+package com.jfrog.ide.common.tree;
+
+public class SeverityReason {
+    String name;
+    String description;
+    boolean isPositive;
```
Comment (yahavi) [inline src/main/java/com/jfrog/ide/common/tree/SeverityReason.java:6]:
```suggestion
    private final String name;
    private final String description;
    private final boolean positive;
```

File: src/main/java/com/jfrog/ide/common/tree/VulnerabilityOrViolation.java [hunk 16]
```diff
@@ -0,0 +1,19 @@
+package com.jfrog.ide.common.tree;
+
+import javax.swing.tree.DefaultMutableTreeNode;
+
+public abstract class VulnerabilityOrViolation extends DefaultMutableTreeNode implements SubtitledTreeNode {
```
Comment (yahavi) [inline src/main/java/com/jfrog/ide/common/tree/VulnerabilityOrViolation.java:5]:
```suggestion
public abstract class VulnerabilityOrViolationNode extends DefaultMutableTreeNode implements SubtitledTreeNode {
```

File: src/main/java/com/jfrog/ide/common/tree/VulnerabilityOrViolation.java [hunk 17]
```diff
@@ -0,0 +1,19 @@
...
+    public abstract Severity getSeverity();
+
+    public String getIcon() {
+        return getSeverity().getIconName();
+    }
+
+    public Artifact getParentArtifact() {
+        Object parent = getParent();
```
Comment (yahavi) [inline src/main/java/com/jfrog/ide/common/tree/VulnerabilityOrViolation.java:13]: Is it a bad practice to use "Object" as a type
```suggestion
        TreeNode parent = getParent();
```

File: src/main/java/com/jfrog/ide/common/tree/VulnerabilityOrViolation.java [hunk 18]
```diff
@@ -0,0 +1,19 @@
...
+        if (parent == null || !(parent instanceof Artifact)) {
```
Comment (yahavi) [inline src/main/java/com/jfrog/ide/common/tree/VulnerabilityOrViolation.java:14]: Redundant condition
```suggestion
        if (!(parent instanceof Artifact)) {
```

**Review by yahavi — CHANGES_REQUESTED** [review]
Well done @asafgabai!
Please see my inline comments and also make sure to get rid of all compiler warnings.

File: src/main/java/com/jfrog/ide/common/tree/BaseTreeNode.java [hunk 19]
```diff
@@ -0,0 +1,9 @@
+package com.jfrog.ide.common.tree;
+
+import javax.swing.tree.DefaultMutableTreeNode;
+
+public class BaseTreeNode extends DefaultMutableTreeNode {
+    public void sortChildren() {
+        children.sort((treeNode1, treeNode2) -> ((FileTreeNode) treeNode2).getTopSeverity().ordinal() - ((FileTreeNode) treeNode1).getTopSeverity().ordinal());
+    }
+}
```
[the remark this answers is not in the document]
Comment (asafgabai) [reply src/main/java/com/jfrog/ide/common/tree/BaseTreeNode.java:9]: It's used in jfrog-idea-plugin. I moved it there.

File: src/main/java/com/jfrog/ide/common/tree/DescriptorFileTreeNode.java (hunk 5 above)
Comment (asafgabai) [reply src/main/java/com/jfrog/ide/common/tree/DescriptorFileTreeNode.java:21]: That's a good question.
addDependencies is used only in SingleDescriptorScanManager, so there's one descriptor and all of the Artifacts are its children.
On the other hand, addDependency is used only in MavenScanManager, where there might be more than one descriptor, and each artifact can be a child of more than one of them, so it must be cloned.
I'll move the cloning to MavenScanManager and add comments to make it all clearer.

File: src/main/java/com/jfrog/ide/common/tree/Artifact.java [hunk 20]
```diff
@@ -0,0 +1,104 @@
...
+import javax.swing.tree.DefaultMutableTreeNode;
+import javax.swing.tree.TreeNode;
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * @author yahavi
+ */
+public class Artifact extends DefaultMutableTreeNode implements Serializable, SubtitledTreeNode {
```
[the remark this answers is not in the document]
Comment (asafgabai) [reply src/main/java/com/jfrog/ide/common/tree/Artifact.java:12]: Changed to DependencyNode

File: src/main/java/com/jfrog/ide/common/tree/DescriptorFileTreeNode.java [hunk 21]
```diff
@@ -0,0 +1,50 @@
...
+            }
+        }
+        sortChildren();
+    }
+
+    private void sortChildren() {
+        if (CollectionUtils.isNotEmpty(children)) {
+            children.sort((treeNode1, treeNode2) -> ((Artifact) treeNode2).getTopSeverity().ordinal() - ((Artifact) treeNode1).getTopSeverity().ordinal());
+        }
+    }
```
Comment (yahavi) [inline src/main/java/com/jfrog/ide/common/tree/DescriptorFileTreeNode.java:44]: Can be removed if FileTreeNode would extend BaseTreeNode (and move the BaseTreeNode from jfrog-idea-plugin)
```suggestion
```

Comment (asafgabai) [reply src/main/java/com/jfrog/ide/common/tree/DescriptorFileTreeNode.java]: This logic is changed in https://github.com/jfrog/ide-plugins-common/pull/97, so it's not relevant anymore.

**Comment by asafgabai** [comment]
[שינוי UI](https://trello.com/c/Qbw43Czs/61-%D7%A9%D7%99%D7%A0%D7%95%D7%99-ui)

**Pull request merged**